programs/cli_wallet/main.cpp

```diff
@@ -276,43 +257,93 @@ int main( int argc, char** argv )
 
       if( !options.count( "daemon" ) )
       {
-         wallet_cli->register_api( wapi );
-         wallet_cli->start();
+         auto wallet_cli = std::make_shared<fc::rpc::cli>( GRAPHENE_MAX_NESTED_OBJECTS );
+         for( auto& name_formatter : wapiptr->get_result_formatters() )
+            wallet_cli->format_result( name_formatter.first, name_formatter.second );
 
-         fc::set_signal_handler([](int signal) {
-            ilog( "Captured SIGINT not in daemon mode" );
-            fclose(stdin);
-         }, SIGINT);
-
-         fc::set_signal_handler([](int signal) {
-            ilog( "Captured SIGTERM not in daemon mode" );
-            fclose(stdin);
-         }, SIGTERM);
+         if( wapiptr->is_new() )
+         {
+            std::cout << "Please use the set_password method to initialize a new wallet before continuing\n";
+            wallet_cli->set_prompt( "new >>> " );
+         }
+         else
+            wallet_cli->set_prompt( "locked >>> " );
+
+         boost::signals2::scoped_connection locked_connection( wapiptr->lock_changed.connect(
+            [wallet_cli](bool locked) {
+               wallet_cli->set_prompt( locked ? "locked >>> " : "unlocked >>> " );
+            }));
+
+         auto sig_set = fc::set_signal_handler( [wallet_cli](int signal) {
+            ilog( "Captured SIGINT not in daemon mode, exiting" );
+            fc::set_signal_handler( [](int sig) {}, SIGINT ); // reinstall an empty SIGINT handler
+            fc::usleep( fc::milliseconds(100) ); // sleep a while to cleanup things
+            wallet_cli->cancel();
```

Comment:
Here exists a race condition.

When a user pressed CTRL+C, editline will signal the program a SIGINT, then break out from `readline()`. The SIGINT will be caught here and be processed in a signal-handling thread. In the meanwhile, the CLI thread continues running, and may call `readline()` again in `get_line` thread wait for user input via `read()` in next loop. The `wallet_cli->cancel()` call here will send another SIGINT to the `get_line` thread.
* if the CLI thread and the get_line thread are fast enough, it would be blocking at `read()` when `wallet_cli->cancel()` executes, thus `read()` will be interrupted, then the program will end happily;
* if the CLI thread or the get_line thread is too slow, `wallet_cli->cancel()` executes before `read()`, the program will get stuck.

The `sleep` here is to increase the chance of not getting stuck. But sometimes it still happens.

Actually, setting `cli_quitting` to `true` in `cancel()` before sending SIGINT to `get_line` thread would help a bit, although it will create a chance so the `readline` job may be already done when the `get_line` thread received the SIGINT, in this case the program will crash. Anyway, we're talking about quitting CLI, so crash is better than hang. I will create a new PR in FC.

Reply:
It seems setting `cli_quitting` to `true` in `cli::cancel()` does little help, the program still hangs sometimes.

It does a little help though: before the change, when the program is stuck, it will quit only on ENTER key; after the change, it will quit on any key.

Reply:
Looks like we can remove the `sleep` if have `cli_quitting = true` in `cli::cancel()`.